src/observer.ts

```diff
@@ -12,50 +14,57 @@ export function observer<P>(baseComponent: FunctionComponent<P>): FunctionCompon
     if (isUsingStaticRendering) {
         return baseComponent
     }
+
     const baseComponentName = baseComponent.displayName || baseComponent.name
+
     // memo; we are not intested in deep updates
     // in props; we assume that if deep objects are changed,
     // this is in observables, which would have been tracked anyway
-
-    const memoComponent = memo(props => {
-        // forceUpdate 2.0
-        const forceUpdate = useForceUpdate()
-
-        // create a Reaction once, and memoize it
-        const reaction = useMemo(
-            () =>
-                // If the Reaction detects a change in dependency,
-                // force a new render
-                new Reaction(
-                    `observer(${baseComponentName})`,
-                    forceUpdate
-                ),
-            []
-        )
-
-        // clean up the reaction if this component is unMount
-        useUnmount(() => reaction.dispose())
+    const memoComponent = memo((props: P) => {
+        const observerReaction = useObserverReaction(baseComponentName)
 
         // render the original component, but have the
         // reaction track the observables, so that rendering
         // can be invalidated (see above) once a dependency changes
-        let rendering
-        reaction.track(() => {
-            rendering = baseComponent(props as P)
+        let rendering!: React.ReactElement<any> | null
+        observerReaction.track(() => {
+            rendering = baseComponent(props)
         })
         return rendering
     })
     memoComponent.displayName = baseComponentName
-    return memoComponent
+    return memoComponent as any
```

Comment:
the type of memoized component is not compatible with FunctionComponent since memoized components allow forward references to the inner component by default
I think there should be a test to make sure forward references work when using observer and that their typing is ok actually

Reply:
I see. That's probably more of the issue of React typings and it feels to be omitted like this, but I guess it's ok as long as it doesn't break the outer contract.